src/Http/Routing/src/Matching/DfaMatcherBuilder.cs

```diff
@@ -117,39 +93,66 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
             //
             // We'll sort the matches again later using the *real* comparer once building the
             // precedence part of the DFA is over.
-            var precedenceDigitComparer = Comparer<(RouteEndpoint endpoint, int precedenceDigit, List<DfaNode> parents)>.Create((x, y) =>
+            var precedenceDigitComparer = Comparer<DfaBuilderWorkerWorkItem>.Create((x, y) =>
             {
                 return x.precedenceDigit.CompareTo(y.precedenceDigit);
             });
 
+            var dfaWorker = new DfaBuilderWorker(work, precedenceDigitComparer, includeLabel, _parameterPolicyFactory);
+
             // Now we process the entries a level at a time.
             for (var depth = 0; depth <= maxDepth; depth++)
+            {
+                dfaWorker.ProcessLevel(depth);
+            }
+
+            // Build the trees of policy nodes (like HTTP methods). Post-order traversal
+            // means that we won't have infinite recursion.
+            root.Visit(ApplyPolicies);
+
+            return root;
+        }
+
+        private struct DfaBuilderWorker
```

Comment:
IMO this should be a class.
- It's quite a large struct
- It's mutable
- Only one is created

Reply:
fair enough